http-server-netty/src/main/java/io/micronaut/http/server/netty/converters/NettyConvertersSpi.java

```diff
@@ -162,7 +162,9 @@ private TypeConverter<Attribute, CompletedPart> attributeToCompletedPartConverte
                     return Optional.empty();
                 }
 
-                return Optional.of(new NettyCompletedAttribute(object));
+                // converter does not claim the input object, so we need to retain it here. it's
+                // released by NettyCompletedAttribute.get*
+                return Optional.of(new NettyCompletedAttribute(object.retain()));
```

Comment:
Why is this `retain` now required, when previously it wasn't? I can't see a change to NettyCompletedAttribute to handle this?

Reply:
i think it was a bug that never presented itself, probably because conversion was happening early enough. the new behavior is correct, because it is consistent with other converters. e.g. httpDataToStringConverter does not change the refCnt of the input. but NettyCompletedAttribute wants to release the object, so we need to retain it here

We have code that only releases if refCnt>0, so this kind of bug can be masked easily

but i just saw, the same applies to the fileupload code above, so i will add a retain there too.

Reply:
haha i think changing it in the fileupload code triggered leak detection. so i guess i have to look at this more after all.

Reply:
oh it's because NettyCompletedFileUpload constructor does a retain, but NettyCompletedAttribute does not. brilliant.